packages/util/src/polymorphism.ts

```diff
@@ -0,0 +1,193 @@
+/* eslint-disable @typescript-eslint/no-explicit-any */
+
+/**
+ * NOTE: Polymorphic approach inspired by(lifted from) reach-ui
```

Comment:
Would love this in 1+n commits, if possible:

1 - The "lifted from" state
n - The modifications made to it

Would be great to see what had to change to fit our approach.

Reply:
commit history updated. all code is copied with the exception of the `forwardRefWithAsAndStatics` method with is an addition